sklearn/metrics/classification.py

```diff
@@ -391,72 +392,192 @@ def jaccard_similarity_score(y_true, y_pred, normalize=True,
+        in case `average='samples'`.
 
     sample_weight : array-like of shape = [n_samples], optional
         Sample weights.
 
     Returns
     -------
-    score : float
-        If ``normalize == True``, return the average Jaccard similarity
-        coefficient, else it returns the sum of the Jaccard similarity
-        coefficient over the sample set.
-
-        The best performance is 1 with ``normalize == True`` and the number
-        of samples with ``normalize == False``.
+    score: float (if average is not None) or array of float, shape =\
+            [n_unique_labels]
 
     See also
     --------
     accuracy_score, hamming_loss, zero_one_loss
 
     Notes
     -----
-    In binary and multiclass classification, this function is equivalent
-    to the ``accuracy_score``. It differs in the multilabel classification
-    problem.
+    It differs in implementation from ``accuracy_score`` from all three
+    classifications i.e. binary, mutliclass and multilabel.
 
     References
     ----------
     .. [1] `Wikipedia entry for the Jaccard index
            <https://en.wikipedia.org/wiki/Jaccard_index>`_
 
-
     Examples
     --------
     >>> import numpy as np
     >>> from sklearn.metrics import jaccard_similarity_score
     >>> y_pred = [0, 2, 1, 3]
     >>> y_true = [0, 1, 2, 3]
-    >>> jaccard_similarity_score(y_true, y_pred)
+    >>> jaccard_similarity_score(y_true, y_pred, average='macro')
     0.5
-    >>> jaccard_similarity_score(y_true, y_pred, normalize=False)
-    2
+    >>> jaccard_similarity_score(y_true, y_pred, average='micro')
+    ... # doctest: +ELLIPSIS
+    0.33...
 
     In the multilabel case with binary label indicators:
 
-    >>> jaccard_similarity_score(np.array([[0, 1], [1, 1]]),\
-        np.ones((2, 2)))
+    >>> jaccard_similarity_score(np.array([[0, 1], [1, 1]]),
+    ... np.ones((2, 2)))
     0.75
+
+    In the multiclass case:
+
+    >>> y_pred = ['ant', 'ant', 'cat', 'cat', 'ant', 'cat']
+    >>> y_true = ['cat', 'ant', 'cat', 'cat', 'ant', 'bird']
+    >>> jaccard_similarity_score(y_true, y_pred) # doctest: +ELLIPSIS
+    0.388...
     """
 
+    average_options = (None, 'binary', 'micro', 'macro', 'weighted', 'samples')
+    if average not in average_options:
+        raise ValueError("average has to be one of " + str(average_options))
+
     # Compute accuracy for each possible representation
     y_type, y_true, y_pred = _check_targets(y_true, y_pred)
     check_consistent_length(y_true, y_pred, sample_weight)
+    present_labels = unique_labels(y_true, y_pred)
+
+    if average == 'binary':
+        if y_type == 'binary':
+            if pos_label not in present_labels:
+                if len(present_labels) < 2:
+                    return 0.
+                else:
+                    raise ValueError("pos_label=%r is not a valid label: "
+                                    "%r" % (pos_label, present_labels))
+            labels = [pos_label]
```

Comment:
I am not sure as to what the best way to deal with `y_type = binary` (a superset of `average='binary'`) is.

I can simply replace this line of code with `return accuracy_score(y_true, y_pred)` but the cases of `y_type == 'binary'` and `average != binary` will be dealt with in `y_type == 'multiclass'` case.

Reply:
average='binary' (evaluating just a single class) requires y_type = 'binary', but the opposite is not true. For instance, imagine you have 3 classes in your dataset, but in a particular test portion, you only have two of the three. If you were using a multi-class metric throughout your CV, when presented with only two classes, you would still treat your binary data as if it were multiclass and evaluate each class rather than just the positive class.